### DISCUSSION TITLE
Where is an implementation for Simple Recovery Algorithm?

### DISCUSSION THREAD

**Discussion by ty-bnn** [GumTreeDiff/gumtree#379] [Q&A] [answered]
I am making changes to GumTree Algorithm for my research.
When I run webdiff with no options except file paths, **what kind of recovery algorithm are applied by default**?
According to your paper, Fine-grained, Accurate, and Scalable Source Code Differencing, you have three kind of recovery algorithms, which are optimal, simple and hybrid of them.
And I want to know where the **Simple Recovery Algorithm** is implemented?

**Comment by jrfaller** [comment, marked as answer]
Hi @ty-bnn !

The Simple algorithm is the default one starting from GumTree 4. You can take a look at the implementation of most GumTree's matchers starting in this class : https://github.com/GumTreeDiff/gumtree/blob/main/core/src/main/java/com/github/gumtreediff/matchers/CompositeMatchers.java.

Also, don't hesitate to take a look into our auto matchers that are based upon the work we did on hyper parameter optimization here : https://github.com/GumTreeDiff/gumtree/blob/main/core/src/main/java/com/github/gumtreediff/matchers/AutoMatchers.java.

Cheers!

**Comment by ty-bnn** [comment]
Hi @jrfaller 
Thanks for sharing tips!
When I run GumTree 4, it seems to work `SimpleGumTree` as a matcher as described  [here](https://github.com/GumTreeDiff/gumtree/blob/7c3df6175daa51fd6726db5155d44b18ee0bb0d8/core/src/main/java/com/github/gumtreediff/matchers/CompositeMatchers.java#L81), which means the tool runs `GreedyMatcher` as Top-Down phase and `SimpleBottomUpMathcer` as Bottom-Up and Recovery phase right?

**Reply by jrfaller** [reply]
Exact!

**Reply by ty-bnn** [reply]
I understood.
Thank you!
Have a happy new year!!
